torch/_dynamo/metrics_context.py

```diff
@@ -0,0 +1,82 @@
+from typing import Any, Callable, Dict, Optional, Type
+from typing_extensions import TypeAlias
+
+
+OnExitType: TypeAlias = Callable[[Dict[str, Any]], None]
+
+
+class MetricsContext:
+    def __init__(self, on_exit: OnExitType):
+        """
+        Use this class as a contextmanager to create a context under which to accumulate
+        a set of metrics, e.g., metrics gathered during a compilation. On exit of the
+        contextmanager, call the provided 'on_exit' function and pass a dictionary of
+        all metrics set during the lifetime of the contextmanager.
+        """
+        self._on_exit = on_exit
+        self._metrics: Dict[str, Any] = {}
+        self._level = 0
+
+    def __enter__(self) -> "MetricsContext":
+        """
+        Initialize metrics recording.
+        """
+        if self._level == 0:
+            # In case of recursion, track at the outermost context.
+            self._metrics = {}
+
+        self._level += 1
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        _traceback: Any,
+    ) -> None:
+        """
+        At exit, call the provided on_exit function.
+        """
+        self._level -= 1
+        assert self._level >= 0
+        if self._level == 0:
+            self._on_exit(self._metrics)
+
+    def increment(self, metric: str, value: int) -> None:
+        """
+        Increment a metric by a given amount.
+        """
+        if self._level == 0:
+            raise RuntimeError(f"Cannot increment {metric} outside of a MetricsContext")
```

Comment:
Because of AOTInductor and other "non-standard", currently uninstrumented paths, you have to make this inert when level == 0. Would be OK to follow up and force consistent instrumentation for these paths in the future. This is some of your CI failures.

Reply:
The other case I see in CI is the newly added runtime autotuning, which also is unscoped (it should get scoped, but not in this PR)